FC-1145: :memo:  Foundations Chapter (#8)
Templum/developers-community: pages/Documents/DeveloperTools/LivePersonFunctions/Foundations/features.md

permalink: liveperson-functions-foundations-features.html
indicator: both
redirect_from:
  - liveperson-functions-scheduled-invocations.html
---

### Scheduling

> Some usecases require a scheduled invocation of a function. Two example for this would be OAuth 2.0 related refresh operations or scheduled invocation to keep [functions warm](liveperson-functions-foundations-concepts.html#cold-start). We allow the scheduling of functions based on CRON expressions, we suggest to leverage [this page](https://crontab.guru/) to verify your expression. Please be aware of the following limitation:

Reviewer: ```suggestion
Some usecases require a scheduled invocation of a function. Two examples for this would be OAuth 2.0 related refresh operations or scheduled invocation to keep [functions warm](liveperson-functions-foundations-concepts.html#cold-start). We allow the scheduling of functions based on CRON expressions, we suggest to leverage [this page](https://crontab.guru/) to verify your expression. Please be aware of the following limitation:
```

{: .important}
Scheduled invocations are always performed in **UTC** timezone and therefore might differentiate from your local time. Please keep that in mind when scheduling any function.

* The lowest frequency for scheduled invocations that we allow is once a minute
* Only productive functions can be scheduled
* Configurable Payload is limited to 8000 characters

{: .notice}
Undeploying a scheduled function will result in the **deletion** of the associated schedule. You will need to **recreate** it, once you redeployed the function.

Schedules can be created by heading to the "Schedule an Invocation"-page reachable via the "Watch"-icon on the sidebar. Here you will also see any existing Schedule. Further it will show you the times for the last and next invocation in your **local** time. Along with an "STATE"-column that will indicate if the last invocation was successful or a failure.

<img class="fancyimage" alt="Functions: Schedules" src="img/functions/functions_schedules_list.png">

> To create a Schedule you will need to hit the "Create a schedule"-Button, than you can choose your productive functions from the drop down. After that provide a valid cron expression. The UI will always calculate the next invocation in your **local** time and display it to you. In case you used invalid expression the UI will provide you with valid example values. Further you may adjust the displayed payload. It will be than used for the scheduled invocation and therefore is available to the function during the invocation.

Reviewer: ```suggestion
To create a Schedule you will need to hit the "Create a schedule"-Button, then you can choose your productive functions from the drop down. After that provide a valid cron expression. The UI will always calculate the next invocation in your **local** time and display it to you. In case you used an invalid expression the UI will provide you with valid example values. Further, you may adjust the displayed payload. It will be then used for the scheduled invocation and therefore is available to the function during the invocation.
```

<img class="fancyimage" alt="Functions: Create Schedule" src="img/functions/functions_schedules_create.png">

### Toolbelt

In order to ease to ease the solution of some common tasks we introduced our runtime library the Toolbelt. It comes with a set of clients and features that allow the accomplishment of a wide variety of tasks. Following clients are included:

* MTLS Client: Allows MTLS based requests to be send, especially relevant for zero-trust environments
* HTTP Client: Allows http/https based requests
* Secret Client: Interaction with our Secret Store during runtime
* Context Service Client: Interaction with Conversation Context Store, right now only supports V1 of the API

> * LP Client: Eases the interactions with common LivePerson Api's

Reviewer: ```suggestion
* LP Client: Eases the interactions with common LivePerson APIs
```

### Secret Store

In this chapter we will highlight some important details around the Secret Store. If you want to learn how to setup a secret or how access within the function works. Please head over to the [configuration page](liveperson-functions-getting-started-configure.html#secrets). You may also make sure your user has the necessary [permissions](liveperson-functions-permission-system.html) to interact with our Secret store.

Our Secret Store allows you to securely store any credentials or secrets you may use for your logic. The creation of secrets happens exclusively via the UI, not the CLI or from a function. Once a secret is created you will no longer be able to see its value in clear-text. Only the key will remain viable.

Modifications of existing secrets is only possible by a function. Please be aware that if a secret is accessed & modified by multiple lambdas, the last writer will win.

Deletion of a secret can only performed via the UI. It's important that with the deletion it becomes unavailable immediately to any functions that is using it. Therefore you should make sure it is not used anywhere prior to deleting the secret.

{: .important}

> During the [provisioning](liveperson-functions-provisioning.html) process of the account a couple of default secrets are created. Those are not displayed by default, but they are viewable.

Reviewer: ```suggestion
During the [provisioning](liveperson-functions-provisioning.html) process of the account, a couple of default secrets are created. Those are not displayed by default, but they are viewable.
```

<img class="fancyimage" alt="Functions: Default Secrets" src="img/functions/functions_secret_defaults.png">

### Domain Whitelisting

> As explained within our [networking foundation](liveperson-functions-foundations-networking.html) the communication between a function and external services is subject to a domain allow list, also called whitelist. The next paragraphs should highlight how to add new domains to the whitelist and also how to remove them. Further it will provide some important hints based on frequent questions received surrounding the whitelisting of domains.

Reviewer: ```suggestion
As explained within our [networking foundation](liveperson-functions-foundations-networking.html), the communication between a function and external services is subject to a domain allow list, also called whitelist. The next paragraphs should highlight how to add new domains to the whitelist and also how to remove them. Further, it will provide some important hints based on frequent questions received surrounding the whitelisting of domains.
```
